### PR TITLE
Suggestions for Strap Community Roadmap

### PR DESCRIPTION
**Pull request by ChristopherA**
As @MikeMcQuaid suggested at #GitHubUniverse 2019 in SF, here are some ideas about building a community around improving strap and related files.

### CODE REVIEW

File: ROADMAP.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+shell script settings for ssh and tmux to VMs).
+
+A Possible Roadmap
+------------------
+
+* [ ] Strap as it stands is very good. My primary suggestiton is that it
+  download the users private repo (if it exists) as .private and run a setup
+  script there. This just require a PR to the current repo and some testing.
+  With this @ChristopherA can continue testing some example dotfiles as best
+  practices.
```
Comment (MikeMcQuaid) [inline ROADMAP.md]: This can be done currently with `script/strap-after-setup`, `CUSTOM_HOMEBREW_TAP` or `CUSTOM_BREW_COMMAND`

File: ROADMAP.md [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+* [ ] At some point when @MikeMcQuaid feels confident that building a community
+  is a good move, move the Strap repo over to a GitHub community, and start
+  referring existing strap users to use that version instead.
```
Comment (MikeMcQuaid) [inline ROADMAP.md]: I'm not sure what a "GitHub Community" means here but I do not have any interest in moving this project to community ownership currently or in future, sorry.

File: ROADMAP.md [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+* [ ] Both @MikeMcQuaid and @ChristopherA invite some other folk to join in the
+  community, and integrate their ideas there. In particular, @thisaaronm has
+  some good stuff in his script, and there are some security experts that maybe
+  can recommend some additional minimal hardening.
```
Comment (MikeMcQuaid) [inline ROADMAP.md]: I am interested in PRs that suggest additional minimal hardening and integrations in general. I'm not sure what are the blockers on that happening as-is, though?

File: ROADMAP.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+* [ ] Work collaboratively to implement best practices for Catalina .zsh
+  on a dotfiles-base, private-base and homebrew-brewfile-base repos. In
+  particular, the use of shell scripts in .d directory will make it easier
+  for people to add their own functionality if they fork these repos.
+
+* [ ] Invite people to contribute opinionated dotfiles-base, private-base and
+  homebrew-brewfile-base for various other purposes, such as other shells,
+  different development environments, etc. With a .d directory architecture, it
+  will be easy to take the best of these and integrate back into -base if the
+  community believes these are best practices for all.
```
Comment (MikeMcQuaid) [inline ROADMAP.md]: This is out of scope for Strap. Having a base for user dot files is a good idea but not something I'm personally interested in working on, integrating in or recommended through Strap.

**Comment by MikeMcQuaid** [comment]
Closing just because this isn't going to be merged as a document but happy to continue discussion on this PR.

File: ROADMAP.md (hunk 1 above)
Comment (ChristopherA) [reply ROADMAP.md]: I've tried this, but `script/strap-after-setup` didn't seem to get the authorization environment that the standup script has.

The goal of .private is to explicitly separate out setting up of private files such as gpg/ssh/tokens, etc. Another goal of private is that there are different approaches to this which your example dotfiles don't support, such as `gitsecrets`, `gitcrypt`, LastPass, etc.

File: ROADMAP.md (hunk 2 above)
Comment (ChristopherA) [reply ROADMAP.md]: I can understand not going to community ownership, the repo would still be yours.

But there are some advantages to leveraging GitHub's community feature. Basically it allows for there to be a master `dotfiles`, `Brewfiles` (and hopefully `private`) repositories for people to fork. Right now we have to submit PRs to your personal repo. The idea is that you'd create github/strap, move your 3 repos there, make Heroku, and accept future PRs there.

Using Github community feature also allows ability to add other people's opinionated different dotfiles that work with Strap for other scenarios. I'm not a web developer so my dotfiles are going to differ significantly from someone who is, but it works with Strap and I'd like to share it. 

It would also allow for some experiments. I'd like to share with others how I setup using .d directories for all the shell scripts or using `gitcrypt`, which may be something that eventually you'd like to see in the master .dotfiles example repo, but would be unlikely to accept until well proven.

File: ROADMAP.md (hunk 3 above)
Comment (ChristopherA) [reply ROADMAP.md]: No blockers, but it would be nice if there was some way to architecturally separate them in your code. Maybe moving to `script/`?

Some suggested hardening features can be irksome. My favorite tool for this (which I can't seem to find right now, it was Yosemite era) pops up an osascript to ask "do you want to turn this on, why and why not" for some of the choices.

But some of the choices are just "do it" and these should probably be turned on by default.

File: ROADMAP.md (hunk 4 above)
Comment (ChristopherA) [reply ROADMAP.md]: Again, having it in the Github community format allows for people to work on them with compatibility with strap, and to use the issues, PRs and wiki there that allow you to separate the conversations with them from your contribution of an initial `dotfiles` which works with strap.

File: ROADMAP.md (hunk 1 above)
Comment (MikeMcQuaid) [reply ROADMAP.md]: I'm not sure what's meant by "authorisation environment". The GitHub/Git credentials are set systemwide and environment variables from this script are intentionally not exported for security reasons.


> The goal of .private is to explicitly separate out setting up of private files such as gpg/ssh/tokens, etc.

That's fine but it's unrelated to Strap.

> Another goal of private is that there are different approaches to this which your example dotfiles don't support, such as `gitsecrets`, `gitcrypt`, LastPass, etc.

My dotfiles are unrelated to Strap and unsupported by me for anyone except me.

File: ROADMAP.md (hunk 2 above)
Comment (MikeMcQuaid) [reply ROADMAP.md]: > But there are some advantages to leveraging GitHub's community feature.

I've used this on many other projects but I don't see Strap as being a good use-case. There has been no-one on Strap that has been involved enough with the project for me to consider adding them as a maintainer and no-one who has submitted merged pull requests to this project who has requested this.

> Basically it allows for there to be a master `dotfiles`, `Brewfiles` (and hopefully `private`) repositories for people to fork.

My `dotfiles` are unrelated to Strap. Strap is usable and useful without `dotfiles`. As a result, I would rather keep these concerns separated entirely. If you wish to build a community collection of `dot files`, `Brewfiles` and `private`: please go ahead.

> The idea is that you'd create github/strap, move your 3 repos there, make Heroku, and accept future PRs there.

Strap is my repository and not GitHub's so I will not move it to the `github` organisation. I have no interest in moving my `dotfiles`.

> It would also allow for some experiments. I'd like to share with others how I setup using .d directories for all the shell scripts or using `gitcrypt`, which may be something that eventually you'd like to see in the master .dotfiles example repo, but would be unlikely to accept until well proven.

You can do this already. I'm not interested in accepting changes to my `dotfiles` that do not make my specifically my life easier. They are open for educational purposes and not for a community.

File: ROADMAP.md (hunk 3 above)
Comment (MikeMcQuaid) [reply ROADMAP.md]: > No blockers

I'd suggest submitting PRs in that case.

> but it would be nice if there was some way to architecturally separate them in your code

I don't think splitting Strap into multiple scripts is worth it.

> But some of the choices are just "do it" and these should probably be turned on by default.

I'm interested in reviewing pull requests that do this if they are submitted. They will only be merged, to warn you, if I'm happy with them being enforced for every Strap user.

File: ROADMAP.md (hunk 4 above)
Comment (MikeMcQuaid) [reply ROADMAP.md]: The only thing that makes dotfiles "compatible" with Strap is having `script/setup` and `script/strap-after-setup` files. I don't think an example repository is useful or necessary for this.

In terms of a general community oriented dotfiles project/organisation/repository: I'm afraid I am not interested in turning my dotfiles into this, maintaining it, integrating it into Strap or linking to it from Strap until it's already a project that exists, is clearly well established and maintained and (probably) would provide some additional value to me personally.
